lib/remote/apilistener.cpp

```diff
@@ -368,7 +368,41 @@ void ApiListener::Stop(bool runtimeDeleted)
 	m_Timer->Stop(true);
 	m_RenewOwnCertTimer->Stop(true);
 
+	m_ListenerWaitGroup->Cancel();
+	m_ListenerWaitGroup->Join();
+
+	DisconnectJsonRpcConnections();
+
+	{
+		std::unique_lock lock(m_HttpClientsLock);
+		Log(LogInformation, "ApiListener")
+			<< m_HttpClients.size() << " HTTP connections remaining.";
+	}
+
+	m_Timer->OnTimerExpired.disconnect_all_slots();
+	m_Timer->OnTimerExpired.connect([this](const Timer *const&) {
+		m_WaitGroup->Cancel();
+		DisconnectHttpClients();
+		m_Timer->Stop();
+	});
+	m_Timer->SetInterval(3);
+	m_Timer->Start();
+
+	/* HTTP API requests can take a long time to process under certain circumstances,
+	 * so unlike with check results and JSON RPC connections we want to cancel the
+	 * WorkGroup right away so it still has time to send out the 503 responses before
+	 * the AsioTlsStream timer closes the socket.
+	 */
+	m_HttpWaitGroup->Cancel();
+	m_HttpWaitGroup->Join();
+
+	/* At this point, everything is disconnected and no new connections,
+	 * JSON or HTTP API calls, check-results etc. should arrive or enter processing.
+	 * Now we wait until everything that is already being processed is done
+	 * or the timeout started above fires.
+	 */
 	m_WaitGroup->Join();
```

Comment:
Isn't that more or less the situation I've described in https://github.com/Icinga/icinga2/pull/10460#discussion_r2125917636?

You do `m_WaitGroup->Cancel()` asynchronously in the background and (more or less) immediately start to `m_WaitGroup->Join()` which could return immediately if there happens to be nothing in flight at that very moment, resulting in this `Stop()` function returning. However, if there may still be messages in flight (which I assume is the case given that this 3 second window exists and `DisconnectJsonRpcConnections()` works asynchronously), if anything in there happens to do `m_WaitGroup->try_shared_lock()` after `Join()` returned but before the 3 second timer fired, `try_shared_lock()` would succeed, it would start new work, but nothing waits for it.